org.eclipse.xtext.xtext.ui.tests/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProviderTest.java

```diff
@@ -232,20 +237,38 @@ private void assertAndApplyAllResolutions(XtextEditor xtextEditor, String issueC
 		IResource resource = xtextEditor.getResource();
 		IMarker[] problems = resource.findMarkers(MarkerTypes.FAST_VALIDATION, true, IResource.DEPTH_INFINITE);
 		assertEquals("Resource should have " + issueCount + " error marker.", issueCount, problems.length);
-
+		validateInEditor(document);
 		MarkerResolutionGenerator instance = injector.getInstance(MarkerResolutionGenerator.class);
 		List<IMarkerResolution> resolutions = Lists.newArrayList(instance.getResolutions(problems[0]));
 		assertEquals(1, resolutions.size());
 		IMarkerResolution resolution = resolutions.iterator().next();
 		assertTrue(resolution instanceof WorkbenchMarkerResolution);
 		WorkbenchMarkerResolution workbenchResolution = (WorkbenchMarkerResolution) resolution;
-		workbenchResolution.run(workbenchResolution.findOtherMarkers(problems), new NullProgressMonitor());
+		IMarker primaryMarker = problems[0];
+		List<IMarker> others = Lists.newArrayList(workbenchResolution.findOtherMarkers(problems));
+		assertFalse(others.contains(primaryMarker));
+		assertEquals(problems.length - 1, others.size());
+		others.add(primaryMarker);
+		workbenchResolution.run(others.toArray(new IMarker[others.size()]), new NullProgressMonitor());
+		if (xtextEditor.isDirty()) {
+			xtextEditor.doSave(null);
+		}
 		InternalBuilderTest.cleanBuild();
 		problems = resource.findMarkers(MarkerTypes.FAST_VALIDATION, true, IResource.DEPTH_INFINITE);
 		assertEquals("Resource should have no error marker.", 0, problems.length);
 
 	}
 
+	private void validateInEditor(IXtextDocument document) {
+		try {
+			Job validationJob = ((XtextDocument)document).getValidationJob();
+			validationJob.schedule();
+			Job.getJobManager().join(ValidationJob.XTEXT_VALIDATION_FAMILY, new NullProgressMonitor());
+		} catch (OperationCanceledException | InterruptedException e) {
+			e.printStackTrace();
```

Comment:
Here we should throw instead and let the test fail

Reply:
Am not sure if this can happen anyway

Reply:
Neither am I, but if it happens, it would be good to let the test fail rather than logging the stuff and have some obscure follow-up issues.

Reply:
updated, lets see